README.md
# Adonis bull provider

Gives nice [adonis](https://github.com/adonisjs/adonis-framework)-like interface with [bull](https://github.com/OptimalBits/bull) library. 

## Installation

1. `npm i adonis-bull-provider --save`
2. register provider inside `start/app.js`

```javascript
const providers = [
    //...
    'adonis-bull-provider/Providers/AdonisBullProvider'
]

const aceProviders = [
    //...
    'adonis-bull-provider/Providers/AdonisBullCommand' // if you want run queue with a command
]
```

3. make `config/redis.js` as per instructions [here](https://adonisjs.com/docs/4.1/redis)

```javascript
module.exports = {
    connection: 'local',
    
    local: {
        host: GET_THIS,
        port: GET_THIS,
        password: GET_THIS,
        db: GET_THIS
    },
}
```
4. add `config/bull.js` configuration *(optional)*

```javascript
module.exports = {
    prefix: 'prefix', // prefix all queues if multiple projects use same database
    defaultRedis: 'local', // as defined in config/redis.js. Can be overridden for each queue. Defaults to local
    queueDirectory: 'app/Queues', //where are your queue files, defaults to this value,
    onBoot: true, // to determine whether to automatically register queues or not
    arenaPrefix: '', // prefix for bull-arena if you include it
    arenaUser: 'admin', // used for auth on bull-arena server
    arenaPassword: '' // used for auth on bull-arena server
    arenaPort: 1212 // port used to serve bull-arena
}
```
Done!

## Use

Inside queue directory (defaults to `app/Queues` make handlers. handlers look like this

`queueDir/SampleQueue`
```javascript

const BaseQueue = use('Queue/BaseQueue')

class SampleQueue extends BaseQueue {
    
    static get key() {
        return 'queueName'
    }
    
    async eventName(payload) {
        //handle jobs with name 'eventName'
    }
    
    async anotherEventName() {
        //handle jobs with name 'anotherEventName'
    }
    
    async defaultHandle(payload ) { 
        //if method with job name is not fond this is executed
    }
}

module.exports = SampleQueue

```

Besides these, you can also add:
```javascript

    static get redis() {return 'local'} //optional, if you want to use different redis conection
    
    static get config() {return {}} //optional, custom config for queue creation. See this in bull documentation
    
    static createQueue() {
        //NOT RECOMMENDED!!! Override way of creating queue.
        return super.createQueue()
    }
```

When adding job:

```javascript
    const SampleQueue = use('Queue')('SampleQueue') //this returns native bull queue objects
    await SampleQueue.add({data: 'payload'}, {delay: 2000})
    return {}
```

## Starting Queues

Queues are registered automatically on server start by default. But you can include `onBoot: false`  in  `config/bull.js` to disable automatic registration.

If you disable automatic registration, it means you want to use the command option instead.

### Using Adonis command to start queues

Use this command to start the queues:

```sh
adonis queue:listen
```

To also run [bull-arena](https://github.com/bee-queue/arena) alongside, add `--arena`

```sh
adonis queue:listen --arena
```

bull-arena will run on the port you specify with `arenaPort` (defaults to `1212`) in `config/bull.js`.

## Exceptions

To handle errors, define `onError(error, payload)` method on your queues. It'll be called if an error occurs while processing a job

## closeAll

When you register this provider, problem will arise: ace commands will not exit. That happens because redis connections are not closed. For that reason (or for any reason tou might need) you may call closeAll() method.

Inside `start/hooks.js`
```javascript
hooks.before.aceCommand(async () => {
    const {closeAll} = use('Queue/Helpers')
    await closeAll()
})
```

*Note* If you have overridden createQueue method on Queue handlers, this may not work properly

## Thanks

Special thanks to creators of AdonisJs and bull


CHANGELOG.md
<a name="2.0.4"></a>

# 2.0.4 (2020-03-16)


### Bug Fixes

- Handle error on process events

<a name="2.0.3"></a>

# 2.0.3 (2019-09-18)

### Bug Fixes

- Fixed multiple subscribe bug

<a name="2.0.2"></a>

# 2.0.2 (2019-09-17)

### Docs

- Update documentation

<a name="2.0.1"></a>

# 2.0.1 (2019-08-28)

### Bug Fixes

- Minor bug fix

<a name="2.0.0"></a>

# 2.0.0 (2019-08-28)

### Optimization

- Reusing bull queues
- Optimise event handler

### Bug Fixes

- Gracefully handling queues directory not found error

<a name="1.0.0"></a>

# 1.0.0 (2019-07-29)

### Code Style

- Make more use of methods than static functions

<a name="0.4.2"></a>

# 0.4.2 (2019-07-17)

### Docs

- Update documentation

<a name="0.4.1"></a>

# 0.4.1 (2019-07-12)

### Bug Fixes

- Fix `queue.process` bug

<a name="0.4.0"></a>

# 0.4.0 (2019-07-12)

### Optimization

- Handle `eventName` in creating queue

<a name="0.3.0"></a>

# 0.3.0 (2019-06-28)

### Docs

- Update documentation to include ability to override `Queue` key.

<a name="0.2.1"></a>

# 0.2.1 (2019-06-27)

### Bug Fixes

- Remove closing queue after `createQueue` is called.
  
<a name="0.2.0"></a>

# 0.2.0 (2019-06-27)

### Enhancement

- Added a `closeAll` function to close queues.
  
<a name="0.1.0"></a>

# 0.1.0 (2019-06-27)

### Optimization

- Removed `package-lock.json`
  
<a name="0.0.2"></a>

# 0.0.2 (2019-06-27)

### First Release !


src/Helpers.js
const Config = use('Config')
const Helpers = use('Helpers')
const instances = use('Queue/Instances')

let BullHelper = {
    getQueueName(fileName) {
        let prefix = Config.get('bull.prefix')
        let Handler = BullHelper.getQueueHandler(fileName)
        let name = Handler.key || fileName
        return  prefix ? `${prefix}_${name}` : name
    },
    
    getQueueConfig(fileName) {
        let Handler = BullHelper.getQueueHandler(fileName)
        let redisConfig = Config.get(`redis.${Handler.redis}`)
        let config = {redis: redisConfig}
        
        if(Handler.config) {
            Object.assign(config, Handler.config)
        }
        
        return config
    },
    
    getQueueDir() {
        let queueDir = Config.get('bull.queueDirectory')
        if(!queueDir) queueDir = Helpers.appRoot('app/Queues')
        else queueDir = Helpers.appRoot(queueDir)
        return queueDir
    },
    
    getQueueHandler(fileName) {
        return require(`${BullHelper.getQueueDir()}/${fileName}`)
    },
    
    async closeAll() {
        for(let queueName of Object.keys(instances)) {
            await instances[queueName].close()
        }
    }
}

module.exports = BullHelper


src/BaseQueue.js
const bull = require("bull");

const { getQueueName, getQueueConfig } = use("Queue/Helpers");
const Config = use("Config");
const instances = use("Queue/Instances");
const Logger = use("Logger");

class BaseQueue {
  static get redis() {
    return Config.get("bull.defaultRedis", "local");
  }

  static createQueue() {
    let queue = this.getQueueInstance();
    let instance = new this();

    queue.process("*", async payload => {
      Logger.info("Received new job %s", payload.name);
      try {
        if (payload.name && instance[payload.name])
          return await instance[payload.name](payload);
        if (instance.defaultHandle)
          return await instance.defaultHandle(payload);
      } catch (e) {
        try {
          if (instance.onError) return await instance.onError(e, payload);
        } catch (err) {
          Logger.error("Unhandled error on handling bool error");
          Logger.error("%o", err);
        }
        Logger.error("Unhandled error on bool event");
        Logger.error("%o", e);
      }
    });

    return queue;
  }

  static getQueueInstance() {
    let name = getQueueName(this.fileName);

    if (!instances[name]) {
      let config = getQueueConfig(this.fileName);
      instances[name] = new bull(name, config);
    }
    return instances[name];
  }
}

module.exports = BaseQueue;


Providers/AdonisBullProvider.js
"use strict";
const { ServiceProvider } = use("@adonisjs/fold");
const fs = require("fs-extra");

class AdonisBullProvider extends ServiceProvider {
  register() {
    this.app.bind("Queue/BaseQueue", () => {
      return require("../src/BaseQueue");
    });
    this.app.bind("Queue/Helpers", () => {
      return require("../src/Helpers");
    });
    this.app.bind("Queue", () => {
      return require("../src/Service");
    });
    this.app.bind("Queue/Instances", () => {
      return require("../src/QueueInstances");
    });
  }

  async boot() {
    const Config = use("Config");
    const { getQueueDir, getQueueHandler } = use("Queue/Helpers");

    let redisConfig = Config.get("redis");
    if (!redisConfig) {
      throw {
        status: 500,
        message:
          "Bull uses redis in background. Make config/redis.js file https://adonisjs.com/docs/4.1/redis"
      };
    }

    try {
      var queues = await fs.readdir(getQueueDir());
    } catch (e) {
      return; //directory does not exist
    }

    let onBoot = Config.get("bull").onBoot;
    if (onBoot == undefined || onBoot == null) onBoot = true;

    const BaseQueue = use("Queue/BaseQueue");
    for (let queue of queues) {
      if (!queue.endsWith(".js")) continue;
      let Handler = getQueueHandler(queue);
      let debugInstance = new Handler();
      if (!(debugInstance instanceof BaseQueue))
        throw {
          status: 500,
          message: `${Handler.name} does not extend BaseQueue class. use it with Queue/BaseQueue and extend it`
        };

      Handler.fileName = queue;
      if (onBoot) Handler.createQueue();
    }
  }
}

module.exports = AdonisBullProvider;


src/Commands/Listen.js
"use strict";

const fs = require("fs-extra");
const Arena = require("bull-arena");
const basicAuth = require("express-basic-auth");

const { Command } = require("@adonisjs/ace");
const { getQueueDir, getQueueHandler } = use("Queue/Helpers");
const Config = use("Config");
const Logger = use("Logger");

class Listen extends Command {
  static get signature() {
    return `
    queue:listen
    { --arena : Run bull arena dashboard }
    `;
  }

  static get description() {
    return "Start the Queue";
  }

  ui(queues_, redisConfig, bullConfig) {
    const { connection } = redisConfig;

    const queues = queues_.map(queue => {
      const handler = getQueueHandler(queue);
      return {
        name: handler.key,
        hostId: handler.fileName.replace(".js", ""),
        host: redisConfig[connection].host,
        port: redisConfig[connection].port || 6379,
        password: redisConfig[connection].password || null,
        type: "bull",
        prefix: redisConfig.keyPrefix || "bull"
      };
    });

    const port = bullConfig.arenaPort || 1212;
    const arenaConfig = Arena(
      {
        queues
      },
      {
        basePath: bullConfig.arenaPrefix || "/",
        disableListen: true,
        useCdn: false,
        port
      }
    );

    const express = require("express");
    const app = express();

    const basicAuthConfig = { challenge: true, users: {} };
    basicAuthConfig.users[bullConfig.arenaUser || "admin"] =
      bullConfig.arenaPassword || "";

    app.use(basicAuth(basicAuthConfig));

    // Make arena's resources (js/css deps) available at the base app route
    app.use("/", arenaConfig);

    const server = app.listen(port, () => {
      Logger.info("Bull arena is listening at: %s", port);
    });

    const shutdown = () => {
      server.close(() => {
        Logger.info("Stopping bull board server");
        process.exit(0);
      });
    };

    process.on("SIGTERM", shutdown);
    process.on("SIGINT", shutdown);
  }

  bullListener(queues) {
    const BaseQueue = use("Queue/BaseQueue");
    for (let queue of queues) {
      if (!queue.endsWith(".js")) continue;
      let Handler = getQueueHandler(queue);
      let debugInstance = new Handler();
      if (!(debugInstance instanceof BaseQueue))
        throw {
          status: 500,
          message: `${Handler.name} does not extend BaseQueue class. use it with Queue/BaseQueue and extend it`
        };

      Handler.fileName = queue;
      Handler.createQueue();
    }
  }

  async handle(args, { arena }) {
    const bullConfig = Config.get("bull");
    let onBoot = bullConfig.onBoot;
    if (onBoot == undefined || onBoot == null) onBoot = true;
    if (onBoot)
      throw {
        status: 500,
        message: "You cannot use the command while onBoot=true"
      };

    const redisConfig = Config.get("redis");
    if (!redisConfig) {
      throw {
        status: 500,
        message:
          "Bull uses redis in background. Make config/redis.js file https://adonisjs.com/docs/4.1/redis"
      };
    }

    try {
      var queues = await fs.readdir(getQueueDir());
    } catch (e) {
      Logger.error("Could not load queue directory %s");
      return; //directory does not exist
    }

    this.bullListener(queues);
    Logger.info("Bull is listening");
    if (arena) {
      this.ui(queues, redisConfig, bullConfig);
    }
  }
}

module.exports = Listen;
